src/sensors.c

```diff
@@ -76,145 +211,157 @@ static bool update_imu(void)
-      static vector_t acc_sum  = { 0.0f, 0.0f, 0.0f };
-      static float acc_temp_sum = 0.0f;
-
-      acc_sum.x += _accel.x;
-      acc_sum.y += _accel.y;
-      acc_sum.z += _accel.z + 9.80665f;
-      acc_temp_sum += _imu_temperature;
-      acc_count++;
-
-      if (acc_count > 1000)
-      {
-        set_param_float(PARAM_ACC_X_BIAS,
-                              (acc_sum.x - get_param_float(PARAM_ACC_X_TEMP_COMP) * acc_temp_sum) / (float)acc_count);
-        set_param_float(PARAM_ACC_Y_BIAS,
-                              (acc_sum.y - get_param_float(PARAM_ACC_Y_TEMP_COMP) * acc_temp_sum) / (float)acc_count);
-        set_param_float(PARAM_ACC_Z_BIAS,
-                              (acc_sum.z - get_param_float(PARAM_ACC_Z_TEMP_COMP) * acc_temp_sum) / (float)acc_count);
-
-        acc_count = 0;
-        acc_sum.x = 0.0f;
-        acc_sum.y = 0.0f;
-        acc_sum.z = 0.0f;
-        acc_temp_sum = 0.0f;
-        calib_acc = false;
-        // we could do some sanity checking here if we wanted to.
-      }
-    }
-
-    if (calib_gyro)
-    {
-      static uint16_t gyro_count = 0;
-      static vector_t gyro_sum = { 0.0f, 0.0f, 0.0f };
-
-      gyro_sum.x += _gyro.x;
-      gyro_sum.y += _gyro.y;
-      gyro_sum.z += _gyro.z;
-      gyro_count++;
-
-      if (gyro_count > 1000)
-      {
-        set_param_float(PARAM_GYRO_X_BIAS, gyro_sum.x / (float)gyro_count);
-        set_param_float(PARAM_GYRO_Y_BIAS, gyro_sum.y / (float)gyro_count);
-        set_param_float(PARAM_GYRO_Z_BIAS, gyro_sum.z / (float)gyro_count);
-
-        gyro_count = 0;
-        gyro_sum.x = 0.0f;
-        gyro_sum.y = 0.0f;
-        gyro_sum.z = 0.0f;
-        calib_gyro = false;
-        // we could do some sanity checking here if we wanted to.
-      }
-    }
-
-    // correct according to known biases and temperature compensation
-    _accel.x -= get_param_float(PARAM_ACC_X_TEMP_COMP)*_imu_temperature + get_param_float(PARAM_ACC_X_BIAS);
-    _accel.y -= get_param_float(PARAM_ACC_Y_TEMP_COMP)*_imu_temperature + get_param_float(PARAM_ACC_Y_BIAS);
-    _accel.z -= get_param_float(PARAM_ACC_Z_TEMP_COMP)*_imu_temperature + get_param_float(PARAM_ACC_Z_BIAS);
+    if (calibrating_acc_flag == true)
+      calibrate_accel();
+    if (calibrating_gyro_flag)
+      calibrate_gyro();
 
-    _gyro.x -= get_param_float(PARAM_GYRO_X_BIAS);
-    _gyro.y -= get_param_float(PARAM_GYRO_Y_BIAS);
-    _gyro.z -= get_param_float(PARAM_GYRO_Z_BIAS);
+    correct_imu();
     return true;
   }
   else
   {
+    // if we have lost 1000 IMU messages something is wrong
+    if(millis() > last_imu_update_ms + 1000)
+    {
+      // change board revision and reset IMU
+      last_imu_update_ms = millis();
+      set_param_int(PARAM_BOARD_REVISION, (get_param_int(PARAM_BOARD_REVISION) >= 4) ? 2 : 5);
+      uint16_t acc1G;
+      mpu6050_init(true, &acc1G, &gyro_scale, get_param_int(PARAM_BOARD_REVISION));
+      accel_scale = 9.80665f/acc1G * get_param_float(PARAM_ACCEL_SCALE);
```

Comment:
The ACCEL_SCALE parameter takes care of the weird MPU6050's that are sometimes on the naze32's. You know, that one that you had for a long time that for some reason reports half of full range.  This routine takes care of that.

I'm not sure how often that happens.  It might be that we could just ignore it, and assume that everyone has a good IMU.